Remove IBM-1047 export option and generate makefiles in IBM-1047 encoding for JDK21 on z/OS

psoujany opened the pull request:
This PR reverts below commit which exports `IBM_JAVA_OPTIONS to IBM-1047`, as this option isn't required anymore. JDK21 z/OS latest builds works fine without any options.
https://github.com/psoujany/TKG/commit/28c7f761d9d2e60043930c1d699ad79c72c2a29c

defaultCharset of Java21 is UTF-8, TKG scripts which are generating makefiles like utils.mk, rerun.mk are getting created in UTF for 21 on z/OS causing encoding issues. Hence, changing scripts to forcefully create these makefiles in `Charset.forName("IBM-1047")` only for 21 on z/OS.

psoujany commented:
Verification runs
z/OS
21 - Grinder_CR/27631/
17 - Grinder_CR/27632
11 - Grinder_CR/27633

zLinux
21 - Grinder_CR/2764

psoujany commented:
@llxia Could you please review this PR. Thank you.

llxia reviewed (changes requested):
This PP does not only revert `IBM_JAVA_OPTIONS="-Dfile.encoding=IBM-1047"`, it sets `JVM_OPTIONS=-Dfile.encoding=IBM-1047` in makefile. If `option isn't required anymore`, then we do not need to set `JVM_OPTIONS=-Dfile.encoding=IBM-1047`

psoujany commented:
> This PP does not only revert `IBM_JAVA_OPTIONS="-Dfile.encoding=IBM-1047"`, it sets `JVM_OPTIONS=-Dfile.encoding=IBM-1047` in makefile. If `this option isn't required anymore. JDK21 z/OS latest builds works fine without any options.`, then we do not need to set `JVM_OPTIONS=-Dfile.encoding=IBM-1047`

The reason we're setting JVM_OPTIONS here, EnvDetector creates autoGenEnv.mk file for 21 it's in UTF-8. Hence added this JVM_OPTIONS. I'll remove this settings in makefile and will add check for 21 in EnvDetector.java itself.

llxia commented:
This PR seems actively being updated. Is it ready for review? If not, please move it to draft.

psoujany commented:
Removed JVM_OPTIONS from makefile and added Charset.forName("IBM-1047") to create autoGenEnv.mk in ebcdic for JDK21+ on z/OS.

Verification
z/OS
21 - Grinder_CR/27875
17 - Grinder_CR/27877
11 - Grinder_CR/27878

zLinux 21 - Grinder_CR/27876

@llxia Could please review the changes. Thank you.

llxia commented on src/org/testKitGen/MkGen.java:
```diff
@@ -251,4 +256,18 @@ private void writeTargets() {
+
+	public Writer getWriterObject(String filetype) {
+		try {
+			if (arg.getSpec().toLowerCase().contains("zos") && (arg.getJdkVersion().matches("2\\d"))) {
```
This regex does not work if we have a Java version 30+.

llxia commented on src/org/testKitGen/UtilsGen.java:
```diff
@@ -15,20 +15,22 @@
 	public UtilsGen(Arguments arg, ModesDictionary md, List<String> ignoreOnRerunList) {
 		this.arg = arg;
 		this.md = md;
+		this.mg = new MkGen(arg, null, null, rerunMk, ignoreOnRerunList, ignoreOnRerunList);
```
- It doesn't seem right to initialize `MkGen` solely to use `getWriterObject()` in `MkGen.java`, especially when it involves setting fake parameters in `MkGen`.

- `getWriterObject()` should be a utility function that is leveraged by all files, including EnvDetector.java.

llxia replied:
Please remove `this.mg = new MkGen(arg, null, null, rerunMk, ignoreOnRerunList, ignoreOnRerunList);`. It is not used.

llxia commented on src/org/testKitGen/ModesDictionary.java:
```diff
@@ -98,7 +98,7 @@ private void parseInvalidSpec(Element modes) throws IOException {
 		int lineNum = 0;
 		BufferedReader reader = null;
-		if (arg.getSpec().toLowerCase().contains("zos")) {
+		if (arg.getSpec().toLowerCase().contains("zos") && !(arg.getJdkVersion().matches("2\\d"))) {
```
Same here. This regex does not work if we have a Java version 30+.

llxia replied:
This has not been addressed

psoujany commented:
Modified scripts as suggested.

Verification jobs
JDK21 zLinux Grinder_CR/27914/ - All Passed

JDK21 z/OS - Grinder_CR/27949/
JDK17 z/OS - Grinder_CR/27950
JDK11 z/OS - Grinder_CR/27951

llxia commented on src/org/testKitGen/UtilsGen.java:
```diff
@@ -40,7 +44,7 @@ public void generateFiles() {
 
 	private void generateUtil() {
-		try (FileWriter f = new FileWriter(utilsMk)) {
+		try (Writer f = Utility.getWriterObject(arg.getJdkVersion(), arg.getSpec(), utilsMk)) {
```
JDK_VERSION is an optional parameter. Please ensure this works if the user does not set JDK_VERSION.

psoujany commented:
@llxia Could you please review the changes. Thank you.

llxia commented on src/org/testKitGen/ModesDictionary.java:
```diff
@@ -98,7 +98,7 @@ private void parseInvalidSpec(Element modes) throws IOException {
 		int lineNum = 0;
 		BufferedReader reader = null;
-		if (arg.getSpec().toLowerCase().contains("zos")) {
+		if (arg.getSpec().toLowerCase().contains("zos") && !(arg.getJdkVersion().matches("[2-9][0-9]"))) {
```
`arg.getJdkVersion()` is still used. Please ensure https://github.com/adoptium/TKG/pull/608#discussion_r1880445022 is addressed.

psoujany commented:
> @psoujany Could you please clarify the purpose of this PR? Based on my observation, the [zos JDK21 sanity.functional test build](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk21_ibm_sanity.functional_s390x_zos_Personal/96/) appears to be fine without this PR.

The main intention of this PR is to run JDK21 tests without using `-Dfile.encoding` options. Once we revert https://github.com/adoptium/TKG/commit/28c7f761d9d2e60043930c1d699ad79c72c2a29c, TKG scripts that generates makefiles like utils.mk, rerun.mk, autoGenEnv.mk will be in ascii for JDK21 due to JEP400(default.Charset() is UTF-8). Hence, the changes proposed in this PR will force test specific makefiles to be in IBM-1047.

llxia commented:
Thanks @psoujany. Some of the Grinder link expired. Could you rerun a non-zos Grinder? Thanks

psoujany commented:
Submitted 11/17/21 jobs on non-z/OS.
11 ppcle - Grinder_CR/28522/
17 zLinux - Grinder_CR/28519/
21 linuxAMD - Grinder_CR/28517/

llxia commented:
[Grinder](https://hyc-runtimes-jenkins.swg-devops.com/view/Test_grinder/job/Grinder_CR/28576/) zos using AQA master branch failed. Is there a related change in AQA repo?

psoujany commented:
@llxia yes, we need to address this issue in AQA. For time being can we fix encoding of jars?
https://github.com/adoptium/aqa-tests/issues/5534

llxia commented:
We need to ensure that the `.gitattributes` zOS update does not impact testing on other platforms.

psoujany commented:
> We need to ensure that the `.gitattributes` zOS update does not impact testing on other platforms.

Raised a PR with the .gitattributes change https://github.com/adoptium/aqa-tests/pull/5833.

llxia commented:
https://github.com/adoptium/aqa-tests/pull/5833 is merged. Please rerun Grinder with this PR and AQA master branch. Thanks

psoujany commented:
Triggered jobs on z/OS 21 with AQA master branch and this PR.
Grinder_CR/29174
Grinder_CR/29175

llxia commented:
[xlinux](https://hyc-runtimes-jenkins.swg-devops.com/view/Test_grinder/job/Grinder_CR/29228/)

llxia reviewed (approved):
LGTM

LongyuZhang reviewed (approved):
LGTM
